src/main.cpp

```diff
@@ -3738,6 +3759,108 @@ bool CVerifyDB::VerifyDB(CCoinsView *coinsview, int nCheckLevel, int nCheckDepth
     return true;
 }
 
+bool RewindBlockIndex(const CChainParams& params)
+{
+    LOCK(cs_main);
+
+    // RewindBlockIndex is called after LoadBlockIndex, so at this point every block
+    // index will have nBranchId set either to the persisted value, or the value of
+    // its parent block. Thus we can reason here that for valid chains:
+    //
+    // - BLOCK_OPT_ACTIVATES_NU must be set only on blocks that activate upgrades.
+    // - nBranchId for each block must match what we expect.
+    auto sufficientlyValidated = [&params](const CBlockIndex* pindex) {
+        auto consensus = params.GetConsensus();
+        bool fFlagSet = pindex->nStatus & BLOCK_OPT_ACTIVATES_NU;
+        bool fFlagExpected = IsActivationHeight(pindex->nHeight, consensus);
+        return fFlagSet == fFlagExpected && pindex->nBranchId == CurrentEpochBranchId(pindex->nHeight, consensus);
```

Comment:
This seems to return 'false' whenever we are not in an activation height.

Reply:
When we are not at an activation height, `fFlagExpected` will indeed be `false`, so this is checking that `pindex->nStatus` does *not* have `BLOCK_OP_ACTIVATES_NU` set (ie. that `fFlagSet` is also false). If `fFlagSet` were true in this case, it would mean that this block index was previously an upgrade point, but no longer is (ie. we changed node software from a version that recognises the upgrade to one that doesn't, e.g. if the test net activation point was changed).

Reply:
OK, I just didn't parse correctly, the `==` is before the `&&`